### DISCUSSION TITLE
Use 'type' as an attribute name in interface declaration

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [eclipse-langium/langium#737] [Q&A] [answered]
## What I Found

Hi guys, I noticed that `type` is a keyword in langium grammar language and is disallowed to be used as an attribute name in interface declaration, so the following rule will be rejected by langium.
```
interface TypeDef {
    type: Type;
    name: string;
}
```
However, when using the type inference mechanism, the following rule is valid and `type` is allowed to be an attribute name:
```
TypeDef:
    'typedef' type=Type 'as' name=ID;
```
I find that 
1. `type` is actually often used as an attribute name, such as the above example;
2. the explicit type declaration mechanism should be preferred for complicated grammar, as suggested by the document

## My Idea
Therefore, I think that supporting 'type' as an attribute name in interface declaration may be helpful to writing complicated grammar.

It seems that `type` can be used implicitly as an attribute name because of the `Assignment` parse rule and `Action` parse rule
https://github.com/langium/langium/blob/9f587815f9d14f433d22831ac11d3b9f11696d9f/packages/langium/src/grammar/langium-grammar.langium#L131-L132
https://github.com/langium/langium/blob/9f587815f9d14f433d22831ac11d3b9f11696d9f/packages/langium/src/grammar/langium-grammar.langium#L83-L84
where the `FeatureName` includes `type` and many other keywords
https://github.com/langium/langium/blob/9f587815f9d14f433d22831ac11d3b9f11696d9f/packages/langium/src/grammar/langium-grammar.langium#L198-L199
while the `TypeAttribute` rule is more restricted, it only allows `ID` as an attribute name thus excluding `type`
https://github.com/langium/langium/blob/9f587815f9d14f433d22831ac11d3b9f11696d9f/packages/langium/src/grammar/langium-grammar.langium#L24-L25

I don't know if excluding these keywords in `TypeAttribute` is an intentional design, such as due to some issues that I have not realized, but as far as I understand, simply replacing `ID` with `FeatureName` in the `TypeAttribute` rule will support the feature I desire, right?

I could be wrong, any comments on this new feature are welcome.

**Comment by montymxb** [comment, marked as answer]
Hi @luan-xiaokun , this is something we're aware of. I also opened up [an issue about this back in June](https://github.com/langium/langium/issues/561). For now you can preface it with a `^` symbol to allow it, like `^type`. 

There is a more appropriate data type rule to allow additional property names (like type) internally, which you've found here in as well (`FeatureName`). It just needs to be added to handle property names in a PR.

**Reply by luan-xiaokun** [reply]
Thanks for your reply! I didn't find that issue when searching, but I'm glad that there is a workaround.
